History store performance improvements

paberr opened the pull request:
## What's in this pull request?

- [x] Improve database layout of history store, removing unnecessary indices with Blake2bHash keys
- [x] Replace `Vec` return values by `Option` since transaction hashes are assumed to be unique
- [x] Make sure that the `ValidityStore` can always enforce the full validity window (fixes #2688)
- [x] Move hash based indices into separate struct and make `Blockchain` configurable
- [x] Expose config option to the outside
- [x] Properly advertise transaction indexing services on the network

Optionally:
- [x] Build indices lazily
- [x] Remove `LightHistoryStore` if no longer needed

> ⚠️ **This PR changes the database layout. Restarting your client with an existing database will create issues!**

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

paberr committed:
Improve history store indexing to use append-only indices where possible.

paberr committed:
Replace `Vec` return values by `Option` for transactions by hashes.
The validity window enforces that only one transaction with the same `raw_tx_hash` can be included during that time.

paberr committed:
Fixes #2688 by keeping at least one additional batch

paberr committed:
Remove indices from history store

paberr committed:
Move indexing into separate struct

paberr committed:
Expose transaction indexing to outside (network/config)

paberr committed:
Improve database performance for indexing

paberr commented:
The optional TODOs will be addressed in a separate PR for readability.

paberr committed:
Implement functionality to rebuild index from scratch

paberr committed:
Rebuild index if turning index on
Fix wasm build and config option

paberr committed:
Benchmark pruning with different methods

paberr committed:
Improve pruning performance (by a factor of 2)

paberr committed:
Use DUP tables for epochs

paberr commented:
## Statistics
Here are some statistics for the improvements of this PR (gathered on a Macbook Pro with M1 CPU).
All statistics are for 720 batches, 700 tpb, rounds: 360, total_txns: 30240000

### Pushing one epoch

| Commit | With Indexing (time) | No Indexing (time) | With Indexing (space) | No Indexing (space) |
| -------- | ------------- | ------------- | ------------- | ------------- |
| **Prior to PR**  | 1259.94s  | 1083.88s | 13017.30Mb  | 9677.65Mb |
| **Improved DB layout but no DUP keys (commit 1ab89269907ca9348ed367b4014a2281acc3d511)** | 162.72s | 116.79s | 8902.33Mb | 6471.89Mb |
| **DUP keys (commit ac390f04d02866963048fd31c1fa8230b5d25ed9)** | 198.85s | 146.07s | 7714.10Mb | 5283.63Mb |

#### Push and commit time per two batches
When looking at the times per "round", we can see that the old performance deteriorates quickly.
![Performance per two batches](https://github.com/nimiq/core-rs-albatross/assets/13586855/d16e93ad-5eb8-4fc8-b78d-6d57d99252b6)

### Pushing multiple epochs

#### Improved DB layout but no DUP keys (commit 1ab89269907ca9348ed367b4014a2281acc3d511)

| Epoch | No Indexing (time) | No Indexing (space) |
| -------- | ------------- | ------------- |
| **1**  | 116.79s | 6471.89Mb |
| **2**  | 131.24s | 12597.94Mb |
| **3**  | 142.02s | 18724.01Mb |

#### DUP keys (commit ac390f04d02866963048fd31c1fa8230b5d25ed9)

| Epoch | No Indexing (time) | No Indexing (space) |
| -------- | ------------- | ------------- |
| **1**  | 146.07s | 5283.63Mb |
| **2**  | 158.99s | 10225.80Mb |
| **3**  | 166.11s | 15167.93Mb |

### Pruning one epoch

| Commit | No Indexing |
| -------- | ------------- |
| **Prior to PR**  | > 2h* |
| **Improved DB layout but no DUP keys (commit 1ab89269907ca9348ed367b4014a2281acc3d511)** | 27.50s |
| **DUP keys (commit ac390f04d02866963048fd31c1fa8230b5d25ed9)** | 0.28s |

_*aborted after passing time of push/commit_
DUP keys already get us a >25,000x improvement

## Rationale for choosing DUP
While non-DUP has faster insertion times (roughly ~30s faster than DUP), the pruning is roughly ~30s slower.
Since insertion times are usually distributed over an epoch (12h) except for the sync, we opted for the DUP approach, which gives us fast pruning times after an epoch. The sync time still improved by an order of magnitude and the DUP approach also results in smaller database sizes.

jsdanielh commented on blockchain/src/history/utils.rs:
```diff
@@ -0,0 +1,157 @@
+    fn from_be_bytes(bytes: &[u8]) -> io::Result<Self> {
+        // It is important to use to_be_bytes here for lexical ordering.
+        Ok(u32::from_be_bytes(
+            bytes.try_into().map_err(|e| io::Error::other(e))?,
```
```suggestion
            bytes.try_into().map_err(io::Error::other)?,
```

jsdanielh commented on pow-migration/src/history/mod.rs:
```diff
@@ -196,7 +196,7 @@ pub async fn migrate_history(
             history_store.add_to_history(
                 &mut txn,
-                0,
+                block_height,
```
I think this won't work. For this the PoW genesis block must be known and this is not the case for the history migration

jsdanielh reviewed (commented):
Haven't finish my review but posting this comments now

jsdanielh commented on consensus/src/sync/light/validity_window.rs:
```diff
@@ -5,6 +5,8 @@ use std::{
 
 use futures::StreamExt;
 #[cfg(feature = "full")]
+use nimiq_blockchain::interface::HistoryInterface;
+#[cfg(feature = "full")]
 use nimiq_blockchain::{Blockchain, CHUNK_SIZE};
```
```suggestion
#[cfg(feature = "full")]
use nimiq_blockchain::{interface::HistoryInterface, Blockchain, CHUNK_SIZE};
```

jsdanielh commented on blockchain/src/history/history_store.rs:
```diff
@@ -31,74 +26,71 @@ use nimiq_transaction::{
         }
     }
 
     /// Gets an historic transaction by its hash. Note that this hash is the leaf hash (see MMRHash)
     /// of the transaction, not a simple Blake2b hash of the transaction.
```
```suggestion
    /// Gets an historic transaction by its epoch number and MMR leaf index.
```

jsdanielh commented on blockchain/src/history/history_store.rs:
```diff
@@ -110,42 +102,136 @@ impl HistoryStore {
+        // Remove all leaves from the history tree and remember the respective hashes and indexes.
+        let num_leaves = tree.num_leaves();
+
+        // Optimisation if we remove all leaves.
```
```suggestion
        // Optimization if we remove all leaves.
```

jsdanielh commented on blockchain/src/history/history_store.rs:
```diff
@@ -158,59 +244,43 @@ impl HistoryStore {
-        txns_size
     }
 
     /// Returns a proof for all the historic transactions at the given positions (leaf indexes). The
```
```suggestion
    /// Returns a proof for all the historic transactions at the given leaf indexes. The
```

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+        let mut txn = self.db.write_transaction();
+        let mut hist_tx_cursor = WriteTransaction::cursor(&txn, &self.history_store.hist_tx_table);
+
+        trace!("Check if history index needs to be rebuilt.");
```
```suggestion
        trace!("Checking if history index needs to be rebuilt.");
```

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+                let tx = tx.get_raw_transaction();
+                addresses
+                    .entry(tx.sender.clone())
+                    .or_insert_with(Vec::new)
```
```suggestion
                    .or_default()
```

viquezclaudio replied:
Clippy warning, use `or_default()`

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+                    .push(ordered_hash.clone());
+                addresses
+                    .entry(tx.recipient.clone())
+                    .or_insert_with(Vec::new)
```
```suggestion
                    .or_default()
```

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+                // We only add reward inherents to the address database.
+                addresses
+                    .entry(ev.reward_address.clone())
+                    .or_insert_with(Vec::new)
```
```suggestion
                    .or_default()
```

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+        }
+    }
+
+    /// Returns a the epoch index and leaf index corresponding to the given
```
```suggestion
    /// Returns the epoch index and leaf index corresponding to the given
```

viquezclaudio replied:
Returns the epoch index and leaf index, corresponding to the given

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+    }
+
+    /// Rebuilds the index from scratch.
+    /// This is a very expensive operation, which currently is only available in an external binary.
```
In an external binary?

jsdanielh commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+        txn.clear_database(&self.tx_hash_table);
+        txn.clear_database(&self.address_table);
+
+        // Iterate over all epochs and leafs.
```
```suggestion
        // Iterate over all epochs and leaves.
```

paberr commented:
Here's a graph of 36 epochs **with indexing** on a different machine. The final database size for this is 270851.91Mb.
![Time per Epoch](https://github.com/nimiq/core-rs-albatross/assets/13586855/0b565ab8-6224-4b81-9a30-f852e481b10c)
We can see that the insertion/commit time stays almost constant.

sisou commented:
Looks like 370-380 seconds per epoch, which is 6.5 minutes. One epoch being 12 hours = 720 minutes, this is a factor of 110. Sounds fine. Plus, this is for _full_ epochs, which we'll likely won't have, so real write will be much faster (per epoch).

Whats a _realistic_ TPB? And how fast would it be then?

viquezclaudio commented on blockchain/src/history/history_store.rs:
```diff
@@ -843,23 +700,36 @@ impl HistoryInterface for HistoryStore {
+        };
 
-        hist_txs
+        self.get_historic_txns(epoch_number, 0..end, Some(txn))
```
There is lots of duplicate code between `get_final_epoch_transactions` and `get_number_final_epoch_transactions`
The main difference is the call to `get_historic_txns` at the end of this function, so maybe it will be worth moving the duplicate code to some common auxiliary function?

viquezclaudio commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+};
+use crate::{history::HistoryTreeChunk, interface::HistoryIndexInterface, HistoryStore};
+
+/// A struct that contains databases to store history trees (which are Merkle Mountain Ranges
```
Might be worth to update the comment :)

viquezclaudio commented on blockchain/src/history/history_store_index.rs:
```diff
@@ -0,0 +1,1445 @@
+    db: DatabaseProxy,
+    /// A database of all epoch numbers and leaf indices indexed by the hash of the (raw) transaction. This way we
+    /// can start with a raw transaction hash and find it in the MMR.
+    /// Mapping of raw tx to epoch number and leaf index.
```
Mapping of raw tx hash to epoch number and leaf index

viquezclaudio commented on blockchain/src/history/utils.rs:
```diff
@@ -0,0 +1,157 @@
+
+/// A transparent wrapper around a value to do fast lookups with `seek_range_subkey`.
+/// It serializes to an empty slice, which is the lower bound.
+/// When deserializing, it will can a different value.
```
Typo in this sentence

viquezclaudio commented on blockchain/src/history/utils.rs:
```diff
@@ -0,0 +1,157 @@
+}
+
+impl<V: FromDatabaseValue> Lookup<V> {
+    /// Only way to initialise a Lookup.
```
Only way to initialize a Lookup.

viquezclaudio commented on blockchain/src/history/utils.rs:
```diff
@@ -0,0 +1,157 @@
+    }
+}
+
+/// A wrapper for an EpochBasedIndex and a Blake2bHash. We use it to for two different functions:
```
The comment mentions two different functions but only one is listed

viquezclaudio commented on blockchain/src/history/validity_store.rs:
```diff
@@ -8,6 +8,7 @@ use nimiq_primitives::policy::Policy;
 /// The validity store is used by full nodes to keep track of which
 /// transactions have occurred within the validity window without
 /// having to store the full transactions
+#[derive(Debug)]
```
Update the comment, it is now used by full nodes and history nodes

viquezclaudio commented on blockchain/src/history/validity_store.rs:
```diff
@@ -123,9 +140,14 @@ impl ValidityStore {
 
-        if num_blocks > Policy::transaction_validity_window_blocks() {
+        // We need to keep at least 'validity_window_blocks' + 'blocks_per_batch' blocks
+        // to account for potential rebranches (when the window moves backwards).
```
It might be worth mentioning this in the validity store documentation:
The validity store keeps at least 'validity_window_blocks' + 'blocks_per_batch' blocks

viquezclaudio reviewed (commented):
Some minor comments, mostly cosmetic

viquezclaudio commented:
Also, I think it is worth favoring DUP keys to get fast prune times,  which are necessary for full nodes

viquezclaudio commented:
I think we should change the name of this PR =)

paberr committed:
Fix table types in index rebuild

viquezclaudio committed:
Add missing validity window tests

viquezclaudio committed:
Fix missing call to validity store

Fix a missing call to validity_store.delete_block_transactions when removing a block from the HistoryStoreIndex

viquezclaudio committed:
Fix issue when computing existing history length

In the validity sync process we need to compute the history tree length (in terms of number of leaves)
to determine the right chunk index that needs to be requested when we already have items in the store (ie. after a restart)
